SproxydKeysScan/DuplicateKeysIngestion.js

```diff
@@ -0,0 +1,279 @@
+const { waterfall } = require('async');
+const { httpRequest } = require('../repairDuplicateVersionsSuite');
+const { SproxydKeysProcessor } = require('./DuplicateKeysWindow');
+const { subscribers } = require('./SproxydKeysSubscribers');
+const { Logger } = require('werelogs');
+const log = new Logger('s3utils:DuplicateKeysIngestion');
+
+const {
+    BUCKETD_HOSTPORT,
+    DUPLICATE_KEYS_WINDOW_SIZE,
+    LOOKBACK_WINDOW,
+} = process.env;
+
+
+/**
+ * @class
+ * @classdesc In order to send keys to DuplicateKeysWindow, we need to read the raft journals.
+ * There is only one of this service per raft session across all servers. Ballot is used
+ * to coordinate the readers
+ */
+class RaftJournalReader {
+    /**
+     *
+     * @param {(number|undefined)} begin - The first cseq number to read from. Automatically set if undefined.
+     * @param {number} limit - How many Raft Journal entries to fetch with each poll.
+     * @param {number} sessionId - Raft Session Id that the reader will poll from.
+     */
+    constructor(begin, limit, sessionId) {
+        this.lookBack = LOOKBACK_WINDOW;
+        this.begin = begin;
+        this.cseq = null;
+        this.limit = limit;
+        this.sessionId = sessionId;
+        this.url = this._getJournalUrl(sessionId);
+        this.processor = new SproxydKeysProcessor(
+            DUPLICATE_KEYS_WINDOW_SIZE,
+            subscribers
+        );
+        this._httpRequest = httpRequest;
+    }
+
+     /**
+     * @param {string} sessionId - Raft session ID
+     * @returns {string} - Url string for given Raft sesssion ID
+     */
+    _getJournalUrl(sessionId) {
+        return `http://${BUCKETD_HOSTPORT}/_/raft_sessions/${sessionId}`;
+    }
+
+    /**
+     * @param {Object} body - Paginated response body from JournalUrl/log
+     * @returns {void}
+     */
+    _setCseq(body) {
+        this.cseq = body.info.cseq;
+    }
+
+    /**
+     * If this.begin is undefined, latest cseq is fetched and this.begin is set to this.cseq - this.lookBack
+     * @param {function(err:Error)} cb - callback
+     * @returns {undefined}
+     * @callback called with (err:Error)
+     */
+    setBegin(cb) {
+        if (this.begin) {
+            return cb(null);
```

Comment:
```suggestion
            return process.nextTick(cb);
```
Because it's safer to always call the callback in a next execution tick, and conventionally we usually write `cb()` in place of `cb(null)` when there is no data to return.